### DISCUSSION TITLE
求和方法

### DISCUSSION THREAD

**Discussion by zhaosu99** [metab0t/PyOptInterface#9] [中文Q&A] [answered]
您好！我对于求和的部分还有些疑问，如果您能不吝赐教，我将非常感激。

第一个问题关于求和方式。我从您的文档中学到了两种求和的方法，第一种是用ExprBuilder，比如

expr = poi.ExprBuilder()
for y in range(100):
    expr += block.μ[y]

这里block是一个SimpleNamespace

第二种是用poi.quicksum。从您的例子中看主要是用在tupledict中。

我的问题是，如果我想求若干变量之和，用哪种方法比较好？

比如 x = [model.add_variable(lb=0) for _ in range(10)]，然后我想限制这些x的和小于等于10。那我应该写：

expr = poi.ExprBuilder()
for i in range(10):
    expr += x[i]
model.add_linear_constraint(expr, poi.Eq, 10)

还是：

model.add_linear_constraint(poi.quick_sum(x), poi.Eq, 10)

第二个问题是quicksum与quicksum_，这两个很相似函数的适用类型分别是什么？他们有什么区别？

谢谢！

**Comment by metab0t** [comment, marked as answer]
可以看一下这里的代码
https://github.com/metab0t/PyOptInterface/blob/e16750acda764c4df98b29bd54ed34df3700f0c8/src/pyoptinterface/_src/aml.py#L72-L86
quicksum本质上就是逐次+=

quicksum_的第一个输入变量表示在这个expr的基础上加和，而quicksum直接返回求和的结果作为一个新的ExprBuilder
